tensorboard/components/vz_line_chart2/line-chart-exporter.ts

```diff
@@ -12,136 +12,126 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-    }
-
-    public exportAsString(): string {
-      const convertedNodes = this.convert(this.root);
-      if (!convertedNodes) return '';
-      const svg = this.createRootSvg();
-      svg.appendChild(convertedNodes);
-      return svg.outerHTML;
-    }
-
-    private createUniqueId(prefix: string): string {
-      return `${prefix}_${this.uniqueId++}`;
-    }
-
-    private getSize(): DOMRect | ClientRect {
-      return this.root.getBoundingClientRect();
-    }
-
-    private createRootSvg(): Element {
-      const svg = document.createElement('svg');
-      const rect = this.getSize();
-
-      // case on `viewBox` is sensitive.
-      svg.setAttributeNS('svg', 'viewBox', `0 0 ${rect.width} ${rect.height}`);
-      svg.setAttribute('xmlns', 'http://www.w3.org/2000/svg');
-      return svg;
-    }
+enum NodeName {
+  GROUP = 'G',
+  DIV = 'DIV',
+  SVG = 'SVG',
+  TEXT = 'TEXT',
+}
 
-    private convert(node: Node): Node | null {
-      let newNode = null;
-      const nodeName = node.nodeName.toUpperCase();
-      if (
-        node.nodeType == Node.ELEMENT_NODE &&
-        (nodeName == NodeName.DIV || nodeName == NodeName.SVG)
-      ) {
-        newNode = document.createElement(NodeName.GROUP);
-        const style = window.getComputedStyle(node as Element);
-        const left = parseInt(style.left, 10);
-        const top = parseInt(style.top, 10);
-        if (left || top) {
-          const clipId = this.createUniqueId('clip');
+export class PlottableExporter {
+  private root: Element;
+  private uniqueId: number = 0;
+  constructor(rootEl: Element) {
+    this.root = rootEl;
+  }
+  public exportAsString(): string {
+    const convertedNodes = this.convert(this.root);
+    if (!convertedNodes) return '';
+    const svg = this.createRootSvg();
+    svg.appendChild(convertedNodes);
+    return svg.outerHTML;
+  }
+  private createUniqueId(prefix: string): string {
+    return `${prefix}_${this.uniqueId++}`;
+  }
+  private getSize(): DOMRect | ClientRect {
+    return this.root.getBoundingClientRect();
+  }
+  private createRootSvg(): Element {
+    const svg = document.createElement('svg');
+    const rect = this.getSize();
+    // case on `viewBox` is sensitive.
+    svg.setAttributeNS('svg', 'viewBox', `0 0 ${rect.width} ${rect.height}`);
+    svg.setAttribute('xmlns', 'http://www.w3.org/2000/svg');
+    return svg;
+  }
+  private convert(node: Node): Node | null {
+    let newNode: Node | null = null;
+    const nodeName = node.nodeName.toUpperCase();
+    if (
+      node.nodeType == Node.ELEMENT_NODE &&
+      (nodeName == NodeName.DIV || nodeName == NodeName.SVG)
+    ) {
+      newNode = document.createElement(NodeName.GROUP);
+      const style = window.getComputedStyle(node as Element);
+      const left = parseInt(style.left || '', 10);
+      const top = parseInt(style.top || '', 10);
+      if (left || top) {
+        const clipId = this.createUniqueId('clip');
+        if (newNode instanceof Element) {
```

Comment:
Aren't we guaranteed `newNode` is always an Element?

Reply:
`newNode` is typed as `Node | null` (L50). That's because (1) `node` in L49 is `Node` and (2) L77 clones `Node`, not `Element`.

Reply:
I agree `newNode` is typed correctly at declaration on L50, but here on L62, we know that it's been set to an element because it passed L56's createElement('G').  So this `if` condition should return true, no matter what.  Would it make sense to cast it to Element inside this if branch?

Reply:
This is one way to do type coercion, too, though.

Reply:
In my opinion, we should not mix type annotations with runtime logic in this way.  Before mixing, I have a guarantee that I can remove unnecessary code, without worrying that I might break the type checker.  Now, we have to be extra careful when rewriting any code we come across, to make distinguish what is actually there for logic vs. type coercion only.